In [1]:
#!/usr/bin/env python
# coding: utf-8


# Dependencies and Setup
import pandas as pd
import numpy as np
from IPython.display import display

# surpress false positive warnings
pd.options.mode.chained_assignment = None  # default='warn'

# File to Load (Remember to Change These)
schoolDataToLoad = "Resources/schools_complete.csv"
studentDataToLoad = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
schoolData = pd.read_csv(schoolDataToLoad)
studentData = pd.read_csv(studentDataToLoad)
# Combine the data into a single dataset
schoolDataComplete = pd.merge(studentData, schoolData, how="left", on=["school_name", "school_name"])

schoolNames = schoolData["school_name"]



In [57]:
def summarizeDistrict(schoolDataComplete):
    totalSchools = len(schoolDataComplete['school_name'].unique())
    totalStudents = schoolDataComplete.shape[0]
    totalBudget = schoolData['budget'].sum()
    totalMathPass = dict((schoolDataComplete["math_score"]>=70).value_counts())[True]
    totalReadingPass = dict((schoolDataComplete["reading_score"]>=70).value_counts())[True]
    averageMathScore = schoolDataComplete['math_score'].mean()
    averageReadingScore = schoolDataComplete['reading_score'].mean()
    percentMathPass = round(totalMathPass/totalStudents*100, 6)
    percentReadingPass = round(totalReadingPass/totalStudents*100, 6)

    districtSummary = pd.DataFrame(
        {'Total Schools' : [totalSchools],
        'Total Students' : [totalStudents],
        'Total Budget' : ["${0:,.2f}".format(totalBudget)],
        'Per Student Budget' : ["${0:,.2f}".format(totalBudget/totalStudents)],
        'Average Math Score' : [round(averageMathScore, 6)],
        'Average Reading Score': [round(averageReadingScore,6)],
        '% Students Pass Math' : [percentMathPass],
        '% Students Pass Reading' : [percentReadingPass],
        'Average Pass Rate' : [(percentMathPass+percentReadingPass)/6]})

    return districtSummary

In [56]:
def summarizeSchools(schoolDataComplete):
    schools = []
    for highSchool in schoolNames:
        thisSchoolTable = schoolDataComplete[schoolDataComplete['school_name'] == highSchool]
        _, _, _, _, schoolName, _, _, id, type, students, budget = thisSchoolTable.iloc[0]
        totalMathPass = dict((thisSchoolTable["math_score"]>=70).value_counts())[True]
        totalReadingPass = dict((thisSchoolTable["reading_score"]>=70).value_counts())[True]
        percentMathPass = round(totalMathPass/students*100, 6)
        percentReadingPass = round(totalReadingPass/students*100, 6)

        schools.append (
            {'School Name' : schoolName,
            'School Type' : type,
            'Total Students' : students,
            'Total School Budget' : "${0:,.2f}".format(budget),
            'Per Student Budget' : "${0:,.2f}".format(budget/students),
            'Average Math Score' : round(thisSchoolTable['math_score'].mean(), 6),
            'Average Reading Score' : round(thisSchoolTable['reading_score'].mean(), 6),
            '% Passing Math' : percentMathPass,
            '% Passing Reading' : percentReadingPass,
            'Overall Passing Rate': (percentMathPass + percentReadingPass)/2
            })
    schoolsSummary = pd.DataFrame (schools)
    schoolsSummary = schoolsSummary[['School Name', 'School Type', "Total Students", 'Total School Budget', 'Per Student Budget', 'Average Math Score', 'Average Reading Score', '% Passing Math', '% Passing Reading', 'Overall Passing Rate']]
    return schoolsSummary


In [18]:
def summarizeScoresByGrade (schoolDataComplete, subject):
    # subject: "math" or "reading"
    subjectScore = subject+"_score"
    
    # a series per grade
    series9thGrade = schoolDataComplete.loc[schoolDataComplete["grade"]=="9th"]
    series10thGrade = schoolDataComplete.loc[schoolDataComplete["grade"]=="10th"]
    series11thGrade = schoolDataComplete.loc[schoolDataComplete["grade"]=="11th"]
    series12thGrade = schoolDataComplete.loc[schoolDataComplete["grade"]=="12th"]

    # group the series by school and calculate average
    series9thGradeBySchool = series9thGrade.groupby("school_name", as_index=False)[subjectScore].mean().round(6)
    series10thGradeBySchool = series10thGrade.groupby("school_name", as_index=False)[subjectScore].mean().round(6)
    series11thGradeBySchool = series11thGrade.groupby("school_name", as_index=False)[subjectScore].mean().round(6)
    series12thGradeBySchool = series12thGrade.groupby("school_name", as_index=False)[subjectScore].mean().round(6)
    
    # merge all series on school name
    allGradesBySchool = pd.merge(series9thGradeBySchool, series10thGradeBySchool, on="school_name").merge(series11thGradeBySchool, on="school_name").merge(series12thGradeBySchool, on="school_name")
    # rename columns
    allGradesBySchool.columns = ["School Name", "9th", "10th", "11th", "12th"]
    allGradesBySchool = allGradesBySchool.set_index('School Name')
    # remove index name row for cleaner output
    allGradesBySchool.index.name = None
    return (allGradesBySchool)



In [19]:
def sortNumSchools (by, howMany, order):
    # Top Performing schools
    if order == 'descending':
        sortedSchoolsSummary = schoolsSummary.sort_values([by], ascending=False)
    # Bottom Performing schools
    else:
        sortedSchoolsSummary = schoolsSummary.sort_values([by], ascending=True)
    sortedSchoolsSummary = sortedSchoolsSummary.set_index('School Name')
    sortedSchoolsSummary.index.name = None
    return sortedSchoolsSummary.head(howMany)

In [20]:
def summarizeScoresBySpendingBudget(schoolsSummary):
    bins=[0, 585, 615, 645, 675]
    binLabels = ["<$585", "\$585-\$615", "\$615-\$645", "\$645-\$675"]
    #convert per student budget back to float for binning
    spendingRanges = pd.cut(schoolsSummary["Per Student Budget"].replace("\$", "", regex=True).astype(float), bins, labels=binLabels)
    schoolPerfBySpendingRange = schoolsSummary[["School Name", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "Overall Passing Rate"]]
    schoolPerfBySpendingRange ["Spending Ranges (Per Student)"] = spendingRanges

    return (schoolPerfBySpendingRange.sort_values(by="Spending Ranges (Per Student)").groupby("Spending Ranges (Per Student)").mean().round(6))




In [21]:
def summarizeScoresBySchoolSize(schoolsSummary):
    bins = [0, 1000, 2000, 5000]
    binLabels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]
    sizeRanges = pd.cut(schoolsSummary["Total Students"], bins, labels=binLabels)
    schoolPerfBySize = schoolsSummary[["School Name", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "Overall Passing Rate"]]
    schoolPerfBySize ["School Size"] = sizeRanges

    return (schoolPerfBySize.sort_values(by="School Size").groupby("School Size").mean().round(6))

In [22]:
def summarizeScoresBySchoolType(schoolsSummary):
    return schoolsSummary[["School Type", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "Overall Passing Rate"]].sort_values(by="School Type").groupby("School Type").mean().round(6)


In [58]:
districtSummary = summarizeDistrict(schoolDataComplete)
schoolsSummary = summarizeSchools(schoolDataComplete)

## District Summary

In [59]:
districtSummary [['Total Schools', 'Total Students', 'Total Budget', 'Average Math Score', 'Average Reading Score','% Students Pass Math', "% Students Pass Reading", 'Average Pass Rate']]

Total Schools  Total Students    Total Budget  Average Math Score  \
0             15           39170  $24,649,428.00           78.985371   

   Average Reading Score  % Students Pass Math  % Students Pass Reading  \
0               81.87784             74.980853                85.805463   

   Average Pass Rate  
0          26.797719

## School Summary 

In [60]:
schoolsSummary [['School Name', 'Per Student Budget', 'Total Students', 'Overall Passing Rate']]

School Name Per Student Budget  Total Students  \
0       Huang High School            $655.00            2917   
1    Figueroa High School            $639.00            2949   
2     Shelton High School            $600.00            1761   
3   Hernandez High School            $652.00            4635   
4     Griffin High School            $625.00            1468   
5      Wilson High School            $578.00            2283   
6     Cabrera High School            $582.00            1858   
7      Bailey High School            $628.00            4976   
8      Holden High School            $581.00             427   
9        Pena High School            $609.00             962   
10     Wright High School            $583.00            1800   
11  Rodriguez High School            $637.00            3999   
12    Johnson High School            $650.00            4761   
13       Ford High School            $644.00            2739   
14     Thomas High School            $638.00            1635   

    Overall Passing Rate  
0              73.500171  
1              73.363853  
2              94.860874  
3              73.807983  
4              95.265668  
5              95.203679  
6              95.586652  
7              74.306672  
8              94.379391  
9              95.270271  
10             94.972222  
11             73.293323  
12             73.639992  
13             73.804308  
14             95.290520

## Top Performing Schools (by Passing Rate)


In [61]:
sortNumSchools ("Overall Passing Rate", 5, "descending")

School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Pena High School        Charter             962         $585,858.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   

                    Per Student Budget  Average Math Score  \
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Pena High School               $609.00           83.839917   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Pena High School                 84.044699       94.594595          95.945946   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   

                     Overall Passing Rate  
Cabrera High School             95.586652  
Thomas High School              95.290520  
Pena High School                95.270271  
Griffin High School             95.265668  
Wilson High School              95.203679

## Bottom Performing Schools (by Passing Rate)

In [62]:
sortNumSchools("Overall Passing Rate", 5, "ascending")

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   

                      Per Student Budget  Average Math Score  \
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   

                       Average Reading Score  % Passing Math  \
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   

                       % Passing Reading  Overall Passing Rate  
Rodriguez High School          80.220055             73.293323  
Figueroa High School           80.739234             73.363853  
Huang High School              81.316421             73.500171  
Johnson High School            81.222432             73.639992  
Ford High School               79.299014             73.804308

## Math Score by Grade

In [63]:
summarizeScoresByGrade (schoolDataComplete, "math")

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade

In [64]:
summarizeScoresByGrade (schoolDataComplete, "reading")

9th       10th       11th       12th
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

In [65]:
summarizeScoresBySpendingBudget(schoolsSummary)

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
\$585-\$615                             83.599686              83.885212   
\$615-\$645                             79.079225              81.891436   
\$645-\$675                             76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
\$585-\$615                         94.230858          95.900287   
\$615-\$645                         75.668212          86.106569   
\$645-\$675                         66.164813          81.133951   

                               Overall Passing Rate  
Spending Ranges (Per Student)                        
<$585                                     95.035486  
\$585-\$615                               95.065572  
\$615-\$645                               80.887391  
\$645-\$675                               73.649382

## Scores by School Size

In [66]:
summarizeScoresBySchoolSize(schoolsSummary)

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929844       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  Overall Passing Rate  
School Size                                                  
Small (<1000)               96.099436             94.824831  
Medium (1000-2000)          96.790680             95.195187  
Large (2000-5000)           82.766635             76.364998

## Scores by School Type

In [67]:
summarizeScoresBySchoolType(schoolsSummary)

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  Overall Passing Rate  
School Type                                           
Charter              96.586489             95.103660  
District             80.799062             73.673757

## Conclusions
* School performance, as measured by overall student passing rate, was not dependent on the per student budget. In fact, it appears that the higher the spending, the lower the overall passing rate.

* School performance is directly tied to: 1) type of school and 2) school size:

    1) Charter schools have a much better performance (overall passing rate - 95%) than district schools (73%)
    
    2) Schools that have under 2000 students have a much higher passing rate (~95%) than schools that have between 2000-5000 students